### DISCUSSION TITLE
New hook for sort order

### DISCUSSION THREAD

**Discussion by Prestaworks** [PrestaShop/PrestaShop#33444] [Ideas]
#### Is your feature request related to a problem?

There is no easy way to add new sorting options for products

#### Describe the solution you'd like

I need a hook that allows a module to manipulate the possible sort orders on the product lists.

[a comment by an automated account is not shown]

**Comment by khouloudbelguith** [comment]
Hi @Prestaworks,

Thanks for your report.
Ping @PrestaShop/product-team what do you think?

Thanks!

**Comment by LouiseBonnard** [comment]
Hi @Prestaworks, thanks for your issue! What do you mean by 'manipulate the possible sort orders on the product list'? @PrestaShop/prestashop-core-developers, do you recommend adding another hook?

**Comment by Prestaworks** [comment]
I want to be able to add and remove these options with a hook.
![Screenshot_2020-09-14 Tillbehör](https://user-images.githubusercontent.com/2902272/93102054-57236780-f6ab-11ea-9e1d-1bfebc8b86a8.png)

**Comment by futurowebpl** [comment]
@khouloudbelguith 

it is same task: https://github.com/PrestaShop/PrestaShop/issues/20535

**Comment by khouloudbelguith** [comment]
Hi @futurowebpl,

Thanks for your feedback.
Ping @Prestaworks can check this issue: https://github.com/PrestaShop/PrestaShop/issues/20535 and confirm that are similar?

Thanks!

**Comment by Prestaworks** [comment]
They are related but not similair, the #20535 seems to be related to the option to select what default options are available on different pages. Mine is the possibility to manipulate the options with modules.

**Comment by PierreRambaud** [comment]
You're already able to change how it works:
https://github.com/PrestaShop/ps_facetedsearch/blob/dev/src/Product/SearchProvider.php#L142

**Comment by Prestaworks** [comment]
> 
> 
> You're already able to change how it works:
> https://github.com/PrestaShop/ps_facetedsearch/blob/dev/src/Product/SearchProvider.php#L142

It's a much better idea to add a hook that allows us to change the content of the drop down.

**Comment by MatShir** [comment]
@PierreRambaud WDYT to add a hook resilient solution than overriding/add custom code?

**Comment by PierreRambaud** [comment]
> @PierreRambaud WDYT to add a hook resilient solution than overriding/add custom code?

Not sure it has been designed for. I will add complexity :thinking:

**Comment by Prestaworks** [comment]
> 
> 
> > @PierreRambaud WDYT to add a hook resilient solution than overriding/add custom code?
> 
> Not sure it has been designed for. I will add complexity 🤔

i was thinking something simple like

instead of
`return [
            $sortSalesDesc->setLabel(
                $translator->trans('Best sellers', [], 'Modules.Facetedsearch.Shop')
            ),
            $sortPosAsc->setLabel(
                $translator->trans('Relevance', [], 'Modules.Facetedsearch.Shop')
            ),
            $sortNameAsc->setLabel(
                $translator->trans('Name, A to Z', [], 'Shop.Theme.Catalog')
            ),
            $sortNameDesc->setLabel(
                $translator->trans('Name, Z to A', [], 'Shop.Theme.Catalog')
            ),
            $sortPriceAsc->setLabel(
                $translator->trans('Price, low to high', [], 'Shop.Theme.Catalog')
            ),
            $sortPriceDesc->setLabel(
                $translator->trans('Price, high to low', [], 'Shop.Theme.Catalog')
            ),
        ];`


it could be something like

`$returnArray = [
            $sortSalesDesc->setLabel(
                $translator->trans('Best sellers', [], 'Modules.Facetedsearch.Shop')
            ),
            $sortPosAsc->setLabel(
                $translator->trans('Relevance', [], 'Modules.Facetedsearch.Shop')
            ),
            $sortNameAsc->setLabel(
                $translator->trans('Name, A to Z', [], 'Shop.Theme.Catalog')
            ),
            $sortNameDesc->setLabel(
                $translator->trans('Name, Z to A', [], 'Shop.Theme.Catalog')
            ),
            $sortPriceAsc->setLabel(
                $translator->trans('Price, low to high', [], 'Shop.Theme.Catalog')
            ),
            $sortPriceDesc->setLabel(
                $translator->trans('Price, high to low', [], 'Shop.Theme.Catalog')
            ),
        ];
hook::exec(NAME OF HOOK, &$retrunarray);
return $returnArray;
`

this would allow us to modify the array and add new sort options.

**Comment by PierreRambaud** [comment]
> > > @PierreRambaud WDYT to add a hook resilient solution than overriding/add custom code?
> > 
> > 
> > Not sure it has been designed for. I will add complexity thinking
> 
> i was thinking something simple like
> 
> instead of
> `return [ $sortSalesDesc->setLabel( $translator->trans('Best sellers', [], 'Modules.Facetedsearch.Shop') ), $sortPosAsc->setLabel( $translator->trans('Relevance', [], 'Modules.Facetedsearch.Shop') ), $sortNameAsc->setLabel( $translator->trans('Name, A to Z', [], 'Shop.Theme.Catalog') ), $sortNameDesc->setLabel( $translator->trans('Name, Z to A', [], 'Shop.Theme.Catalog') ), $sortPriceAsc->setLabel( $translator->trans('Price, low to high', [], 'Shop.Theme.Catalog') ), $sortPriceDesc->setLabel( $translator->trans('Price, high to low', [], 'Shop.Theme.Catalog') ), ];`
> 
> it could be something like
> 
> `$returnArray = [ $sortSalesDesc->setLabel( $translator->trans('Best sellers', [], 'Modules.Facetedsearch.Shop') ), $sortPosAsc->setLabel( $translator->trans('Relevance', [], 'Modules.Facetedsearch.Shop') ), $sortNameAsc->setLabel( $translator->trans('Name, A to Z', [], 'Shop.Theme.Catalog') ), $sortNameDesc->setLabel( $translator->trans('Name, Z to A', [], 'Shop.Theme.Catalog') ), $sortPriceAsc->setLabel( $translator->trans('Price, low to high', [], 'Shop.Theme.Catalog') ), $sortPriceDesc->setLabel( $translator->trans('Price, high to low', [], 'Shop.Theme.Catalog') ), ]; hook::exec(NAME OF HOOK, &$retrunarray); return $returnArray; `
> 
> this would allow us to modify the array and add new sort options.

You're talking about the faceted search module, not how the Core is working. Yes, we can add hooks inside the faceted search module. But it will not fix the "issue" for people not using it :sweat_smile:

**Comment by Prestaworks** [comment]
not really sure how the faceted search module got involved in all this :) My ticket for improvement regards the core really. but it would be nice if the faceted search had the same feature as many merchants use it.

**Comment by samilmarekrygula** [comment]
> You're already able to change how it works: https://github.com/PrestaShop/ps_facetedsearch/blob/dev/src/Product/SearchProvider.php#L142

Is it the way to change this by override? Does someone has an idea how to code this without changing the code directly in this file?

**Comment by samilmarekrygula** [comment]
> > You're already able to change how it works: https://github.com/PrestaShop/ps_facetedsearch/blob/dev/src/Product/SearchProvider.php#L142
> 
> Is it the way to change this by override? Does someone has an idea how to code this without changing the code directly in this file?

I have done this, I have wrotten override for hookProductSearchProvider function in Ps_Facetedsearch module + I have extended SearchProvider class.

**Comment by kpodemski** [comment]
Hello everyone. Following our new approach to feature requests, I converted this Issue to the GitHub Discussion so that the community can discuss the best possible ways to handle this problem and maybe, I hope, organize together to work towards fixing it in the core. If you are motivated to do so, we will be happy to help.
